openedx/features/content_tagging/rest_api/v1/tests/test_views.py

```diff
@@ -0,0 +1,368 @@
+"""
+Tests tagging rest api views
+"""
+
+from unittest import mock
+import ddt
+from django.contrib.auth import get_user_model
+from django.urls import reverse
+from rest_framework import status
+from rest_framework.test import APITestCase
+
+from common.djangoapps.student.auth import add_users, update_org_role
+from common.djangoapps.student.roles import CourseCreatorRole, OrgContentCreatorRole
+from organizations.models import Organization
+from openedx.features.content_tagging.models import Taxonomy
+
+User = get_user_model()
+
+
+# ToDo: import from openedx
+class TestTaxonomyViewSetMixin:
+    def setUp(self):
+        super().setUp()
+
+        self.user = User.objects.create(
+            username="user",
+            email="user@example.com",
+        )
+
+        self.staff = User.objects.create(
+            username="staff",
+            email="staff@example.com",
+            is_staff=True,
+        )
+
+
+@ddt.ddt
+class TestTaxonomyViewSet(TestTaxonomyViewSetMixin, APITestCase):
+    def setUp(self):
+        super().setUp()
+
+        self.org = Organization.objects.create(name="OpenedX", short_name="OeX")
+
+        self.user_group_content_creator = User.objects.create(
+            username="user_group_content_creator",
+            email="user_group_content_creator@example.com",
+        )
+        add_users(self.staff, CourseCreatorRole(), self.user_group_content_creator)
+
+        self.user_org_content_creator = User.objects.create(
+            username="user_org_content_creator",
+            email="user_org_content_creator@example.com",
+        )
+        update_org_role(self.staff, OrgContentCreatorRole, self.user, [self.org])
+
+    @ddt.data(
+        (
```

Comment:
@pomegranited 
This isn't looking good.

At first, I tried to have a dict for each test, but it became HUGE (and messy to read). For each method (CREATE, UPDATE, PATCH, DETAIL, DELETE), we have 4 combinations of `DISABLED_COURSE_CREATION` and `ENABLED_CREATOR_GROUP` flags and 5 (maybe 1 more for an Org Member User) resulting in 20-24 test cases. I tried consolidating some tests but am not satisfied with the result: we cannot verify which case fails because now we only have 4 test cases.

Is there a less verbose way to make the combinations of users x flags to have all 20-24 test cases separately?

I'm thinking of [subtests](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests), but you have a better idea!

Reply:
Yeah, it's a lot.. and we do want our tests to remain as readable as possible.

For the `ENABLES_CREATOR_GROUP` flag, I did this:

* Pulled common setUp data and utility methods into a [TestTaxonomyMixin](https://github.com/openedx/edx-platform/pull/32661/files#diff-ca73cded4994ab137e40fc7011cee435f10feaa439af275937a282c5a8acb5b0R10).
* Wrote my "normal situation" tests with [TestRulesTaxonomy](https://github.com/openedx/edx-platform/pull/32661/files#diff-67e5795cdd1a25095de77990b0e78015b2d49761e60600c5fca1ecee63b191d1R19-R20), which [subclasses TestTaxonomyMixin and enables `ENABLED_CREATOR_GROUP`.
* Wrote my "nobody gets to do anything" tests using a [subclass of TestRulesTaxonomy](https://github.com/openedx/edx-platform/pull/32661/files#diff-67e5795cdd1a25095de77990b0e78015b2d49761e60600c5fca1ecee63b191d1R322-R325), which disables `ENABLED_CREATOR_GROUP` and overrides any changed tests that are in TestRulesTaxonomy.
  Pylint doesn't like this (so you have to quiet it), and you also need to be careful with ddt, and make sure your overridden test methods have the exact same calling data as the parent class.